### PR TITLE
add eiou impl

### PR DESCRIPTION
**Pull request by Wwupup**
## Motivation
In paper: [A Systematic IoU-Related Method: Beyond Simplified Regression for Better Localization](https://ieeexplore.ieee.org/document/9429909) of TIP2021, we propose a new Extended-IOU loss, which outperforms the commonly used GIOU and DIOU in object detection.

## Modification
add implementation of the proposed Extended-IOU Loss (EIOU).

## Use cases (Optional)
In config:
`loss_bbox=dict(type='EIoULoss',  loss_weight=1.0, smooth_point=0.1)`

### CODE REVIEW

File: mmdet/models/losses/iou_loss.py [hunk 1]
```diff
@@ -466,3 +520,46 @@ def forward(self,
             avg_factor=avg_factor,
             **kwargs)
         return loss
+
+
+@MODELS.register_module()
+class EIoULoss(nn.Module):
+
+    def __init__(self,
+                 eps=1e-6,
```
Comment (ZwwWayne) [inline mmdet/models/losses/iou_loss.py:529]: Docstring is required.

**Comment by ZwwWayne** [comment]
Hi @Wwupup ,
Thanks for your kind PR. Could you also help to add unit tests?

File: mmdet/models/losses/iou_loss.py (hunk 1 above)
Comment (Wwupup) [reply mmdet/models/losses/iou_loss.py:529]: Hi, I refer to other existing IoULoss implementation such as DIoULoss, this wrapper class do not need to add docstring， and strdoc should be added to the function `eiou_loss()`.

**Comment by Wwupup** [comment]
> Hi @Wwupup , Thanks for your kind PR. Could you also help to add unit tests?

Hi, I have made modifications to `tests/test_models/test_losses/test_loss.py` in my PR, do I need to make any other changes?

File: mmdet/models/losses/iou_loss.py [hunk 2]
```diff
@@ -231,6 +231,60 @@ def ciou_loss(pred, target, eps=1e-7):
     return loss
 
 
+@weighted_loss
+def eiou_loss(pred, target, smooth_point=0.1, eps=1e-7):
+    r"""Implementation of paper 'Extended-IoU Loss: A Systematic IoU-Related
+     Method: Beyond Simplified Regression for Better Localization,
+
+     <https://ieeexplore.ieee.org/abstract/document/9429909> '.
```
Comment (RangiLyu) [inline mmdet/models/losses/iou_loss.py:239]:
```suggestion
    r"""Implementation of paper `Extended-IoU Loss: A Systematic 
    IoU-Related Method: Beyond Simplified Regression for Better 
    Localization <https://ieeexplore.ieee.org/abstract/document/9429909>`_
```

File: mmdet/models/losses/iou_loss.py (hunk 1 above)
Comment (RangiLyu) [reply mmdet/models/losses/iou_loss.py:529]: The loss class should also add docstring. And you can simply copy the function's docstring to the class.

**Comment by RangiLyu** [comment]
> > Hi @Wwupup , Thanks for your kind PR. Could you also help to add unit tests?
> 
> Hi, I have made modifications to `tests/test_models/test_losses/test_loss.py` in my PR, do I need to make any other changes?

I think that's enough.

File: mmdet/models/losses/iou_loss.py (hunk 1 above)
Comment (Wwupup) [reply mmdet/models/losses/iou_loss.py:543]: OK, I have updated the docstring.

**Review by RangiLyu — APPROVED** [review]

File: mmdet/models/losses/iou_loss.py [hunk 3]
```diff
@@ -466,3 +519,61 @@ def forward(self,
...
+    Code is modified from https://github.com//ShiqiYu/libfacedetection.train.
+
+    Args:
+        pred (Tensor): Predicted bboxes of format (x1, y1, x2, y2),
+            shape (n, 4).
+        target (Tensor): Corresponding gt bboxes, shape (n, 4).
+        smooth_point (float): hyperparameter, default is 0.1.
+        eps (float): Eps to avoid log(0).
+    Return:
+        Tensor: Loss tensor.
```
Comment (RangiLyu) [inline mmdet/models/losses/iou_loss.py:539]: The order of the args should be consistent with the __init__ method.

And delete the "Return"

**Review by RangiLyu — CHANGES_REQUESTED** [review]

**Review by RangiLyu — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Comment by OpenMMLab-Miao** [comment]
Hi @Wwupup ！First of all, we want to express our gratitude for your significant PR in the MMDet project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
